src/realm/object-store/c_api/app.cpp

```diff
@@ -385,6 +385,11 @@ RLM_API realm_user_t* realm_app_get_current_user(const realm_app_t* app) noexcep
 RLM_API bool realm_app_get_all_users(const realm_app_t* app, realm_user_t** out_users, size_t capacity, size_t* out_n)
 {
     return wrap_err([&] {
+        const size_t users_size = (*app)->all_users().size();
+        set_array_size(out_n, users_size);
+        if (capacity < users_size)
+            return false;
```

Comment:
I'm not sure we should be returning `false` here - wouldn't that get interpreted as an error by SDKs? Technically, this is not an error and SDKs should compare `out_n` to `capacity` to decide whether to retry the request with a larger buffer.

Reply:
We can return true, but wouldn't true mean that the SDKs expect the array to be populated?

Reply:
How about we rewrite this like so?
```cpp
RLM_API size_t realm_app_get_all_users(const realm_app_t* app, realm_user_t** out_users, size_t capacity)
{
    return wrap_err([&] {
        const auto& users = (*app)->all_users();
        if (users.size() > capacity) {
            return users.size();
        } else if (users.size() <= 0) {
            return 0;
        }

        REALM_ASSERT(out_users);
        OutBuffer<realm_user_t> buf(out_users);
        for (const auto& user : users) {
            buf.emplace(user);
        }

        buf.release(nullptr);
        return buf.size();
    }, -1);
}
```

and we can use it like:
```cpp
realm_user_t* buf[4];
realm_user_t** users = buf;
size_t count = realm_app_get_all_users2(app, buf, 4);
if (count == 0) {
    // empty array
} else if (count > 4) {
    realm_user_t* buf2[count];
    count = realm_app_get_all_users2(app, buf2, count);
}

if (count == -1) {
    // handle error
}

// users now has all the realm_user_t* entries and count is the total number
```

This is what we do in .NET, and the gist is that we return the count of elements instead of setting it in an our parameter, and the value `-1` signifies error.

Reply:
... and I am not a fan of the suggestion made by @fealebenpae

Reply:
Logically it makes sense that if the data did not get copied we return false in the API, at the same time I don't see this as an error condition that is worth signalling to the SDKs.

@nirinchev  @fealebenpae  we can then do this:
1. add extra _outparam_ to these APIs in order to signal if capacity matched
2. leave things as there are, and the SDKs will have to check that `!(capacity < out_n)` before to try to get the last error in case false is returned.

Reply:
Haha that’s what happens when you submit comments in bulk with a review and a bunch of people add comments in the mean time 😁 

Out of curiosity, is there anything obviously wrong with the approach we use in .NET, @jedelbo? I would like to improve our handling there if that’s the case.

Reply:
My opinion on this is to return false if core did not copy anything. Even returning `size_t` is tricky, eventually we might end up in a corner situation in which we copy `std::numeric_limits<size_t>::max` memory and that will be mistakenly taken as an error. 
The only check that can really guarantee that we have copied all and did it correctly is `capacity == out_n`.

Reply:
I am very much against conflating errors with regular flow. It's not an error condition for the SDK to provide a smaller array - it's actually something that we explicitly expect. SDKs likely have default handling for throwing methods where they check the return value and call `realm_get_last_error` and throw it. Adding custom handling there for small arrays is going to be annoying.

Re: `std::numeric_limits<size_t>::max` - if we end up actually having to copy an array of this size, I'm fairly certain something has gone terribly wrong already so what we return there will be of little consequence.

Reply:
Shouldn't there be an `out_n` argument as well?

Reply:
Yes apologies: 
`RLM_API bool realm_app_get_all_users(const realm_app_t* app, realm_user_t** out_users, size_t capacity, size_t* out_n, bool* data_copied);`

Reply:
That looks okay to me. Assuming that it only returns `false` if an actual error returns, it should fit our needs.

Reply:
Yes, no `false` if there is no error. Otherwise, it will return `true` and set data_copied accordingly.